### PR TITLE
M-546 Osprey

### PR DESCRIPTION
**Pull request by Readystorm546**
SO MUCH NEW STUFF JUST TEST MERGE

### CODE REVIEW

**Comment by Nickvr628** [comment]
19 TC or no deal

**Comment by MonsWTF** [comment]
But does it have an ambient revving noise while held?

**Comment by Nickvr628** [comment]
To port over my suggestion from discord to here, the burst count should be much higher, I say 10. It is supposed to be an uncontrollable hail of bullets, not a bursting rifle.

**Comment by Readystorm546** [comment]
With my most recent commit i have upped the bullet count to 4 and made the spread really high,
The cap is 36 so you can shoot about 9 times before it over heats. Then you have to wait.

V8vXt4Anuc

**Comment by drderp3635** [comment]
Emag to disable heat lock, which lets you fire more, but can make it explode if you fire while overheated

**Comment by boodaliboo** [comment]
ADD TO NUKE OPS FOR 50 TC

**Comment by Firewolf34** [comment]
Excellent PR quality

**Comment by Readystorm546** [comment]
> Emag to disable heat lock, which lets you fire more, but can make it explode if you fire while overheated

hahhahahah no....


> ADD TO NUKE OPS FOR 50 TC

Its on nuke oops for 36 tc at the moment.


> Excellent PR quality

Thanks

**Comment by Bibby0110** [comment]
Why not make the heat sensor emaggable?

**Comment by Alek2ander** [comment]
Does it cost 400000 dollars to fire it for 12 seconds?

**Comment by Readystorm546** [comment]
> Why not make the heat sensor emaggable?

People are going to blowup and die from something and the fact of the matter is. That makes unfun rounds when you just die a loose your nukie round.

**Comment by Readystorm546** [comment]
> Does it cost 400000 dollars to fire it for 12 seconds?

Yes

**Comment by nesisknesis** [comment]
will make the inhand and back sprites consistent if this passes testmerge.

also, if this hasn't already been rectified, it's called the M-556 but called the M-762 in the video in the PR.

File: yogstation/code/modules/projectiles/guns/ballistic/minigun.dm [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	desc = "The massive back stash can hold alot of ammo on your back."
+	icon = 'yogstation/icons/obj/guns/minigunosprey.dmi'
+	icon_state = "holstered"
+	item_state = "backpack"
+	lefthand_file = 'icons/mob/inhands/equipment/backpack_lefthand.dmi'
+	righthand_file = 'icons/mob/inhands/equipment/backpack_righthand.dmi'
+	slot_flags = ITEM_SLOT_BACK
+	w_class = WEIGHT_CLASS_HUGE
+	var/obj/item/gun/ballistic/minigunosprey/gun
+	var/armed = 0 //whether the gun is attached, 0 is attached, 1 is the gun is wielded.
```
Comment (Amelia0010) [inline yogstation/code/modules/projectiles/guns/ballistic/minigun.dm]:
```suggestion
	var/armed = FALSE //whether the gun is attached, FALSE is attached, TRUE is the gun is wielded.
```

File: yogstation/code/modules/projectiles/guns/ballistic/minigun.dm [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	var/overheat = 0
+	var/overheat_max = 36
+	var/heat_diffusion = 1
+
+/obj/item/minigunbackpack/Initialize()
+	. = ..()
+	gun = new(src)
+	START_PROCESSING(SSobj, src)
+
+/obj/item/minigunbackpack/Destroy()
```
Comment (Amelia0010) [inline yogstation/code/modules/projectiles/guns/ballistic/minigun.dm:24]: this should probably qdel the gun

File: yogstation/code/modules/projectiles/guns/ballistic/minigun.dm [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	STOP_PROCESSING(SSobj, src)
+	return ..()
+
+/obj/item/minigunbackpack/process()
+	overheat = max(0, overheat - heat_diffusion)
+
+//ATTACK HAND IGNORING PARENT RETURN VALUE
+/obj/item/minigunbackpack/attack_hand(var/mob/living/carbon/user)
+	if(src.loc == user)
```
Comment (Amelia0010) [inline yogstation/code/modules/projectiles/guns/ballistic/minigun.dm]:
```suggestion
	if(loc == user)
```

File: yogstation/code/modules/projectiles/guns/ballistic/minigun.dm [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+		if(!armed)
+			if(user.get_item_by_slot(SLOT_BACK) == src)
+				armed = 1
```
Comment (Amelia0010) [inline yogstation/code/modules/projectiles/guns/ballistic/minigun.dm]:
```suggestion
				armed = TRUE
```

File: yogstation/code/modules/projectiles/guns/ballistic/minigun.dm [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+				if(!user.put_in_hands(gun))
+					armed = 0
```
Comment (Amelia0010) [inline yogstation/code/modules/projectiles/guns/ballistic/minigun.dm]:
```suggestion
					armed = FALSE
```

File: yogstation/code/modules/projectiles/guns/ballistic/minigun.dm [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	if(armed)
+		icon_state = "notholstered"
+	else
+		icon_state = "holstered"
+
+/obj/item/minigunbackpack/proc/attach_gun(var/mob/user)
+	if(!gun)
+		gun = new(src)
+	gun.forceMove(src)
+	armed = 0
```
Comment (Amelia0010) [inline yogstation/code/modules/projectiles/guns/ballistic/minigun.dm]:
```suggestion
	armed = FALSE
```

File: yogstation/code/modules/projectiles/guns/ballistic/minigun.dm [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	if(user)
+		to_chat(user, "<span class='notice'>You attach the [gun.name] to the [name].</span>")
+	else
+		src.visible_message("<span class='warning'>The [gun.name] snaps back onto the [name]!</span>")
```
Comment (Amelia0010) [inline yogstation/code/modules/projectiles/guns/ballistic/minigun.dm]:
```suggestion
		visible_message("<span class='warning'>The [gun.name] snaps back onto the [name]!</span>")
```

**Review by Amelia0010 — CHANGES_REQUESTED** [review]
should be fine after this

**Comment by Firewolf34** [comment]
Is there any way you could make the heat level of the gun visible? I used it yesterday and my principle complaint was not knowing when it would be ready to fire again. It's hard to manage the heat level when you don't know what it is.

Maybe you could examine the gun and it could tell you how hot the gun feels? Or maybe have the icon of the barrel glow red depending on heat level? Something to give some feedback on how close it is to overheating would be nice.

Some more feedback: it's cooling rate did seem relatively acceptable during my test. Personally, I'd like it to just cool down a *tad bit* faster, not a huge amount but a little bit. Considering you are giving up that backpack slot, and are relying on it completely.

But overall, really cool. It won me a nukie round, saving my ass as I went solo against the entire station because my other squadmembers immediately died to space exposure out the gate. Sooooo... it was just me and the Osprey against the world, there.

I like that you can walk faster by stowing it. Nice touch.

**Comment by Firewolf34** [comment]
Oh and one last thing. Is there a reason you chose that TC amount specifically? It's exactly 1 TC more than the starting TC amount for the team leader, at least it was in my case (I know sometimes they scale it based on pop). Perhaps that was intentional, perhaps not, but it was tough to convince my team members to give me that 1 TC. lol

**Comment by Readystorm546** [comment]
> Oh and one last thing. Is there a reason you chose that TC amount specifically? It's exactly 1 TC more than the starting TC amount for the team leader, at least it was in my case (I know sometimes they scale it based on pop). Perhaps that was intentional, perhaps not, but it was tough to convince my team members to give me that 1 TC. lol
Thats why i did it. To promote sharing and caring.

**Comment by Readystorm546** [comment]
> will make the inhand and back sprites consistent if this passes testmerge.
> 
> also, if this hasn't already been rectified, it's called the M-556 but called the M-762 in the video in the PR.

I fixed it thanks for reminding me :)

**Comment by ynot01** [comment]
imagine having a functional L6 SAW with 94TC worth of ammo (+the original price of the gun) for 36TC

**Comment by ynot01** [comment]
just saw the latest commit, sorry, 60TC worth of ammo

**Comment by Readystorm546** [comment]
> just saw the latest commit, sorry, 60TC worth of ammo

Assuming you hit all your shots and you know dont get shot in the process or overheat or get caught due to slowdown and you are willing to lose your backpack slot and give up all your TC +1 on a non war round for one item. Now i understand if its war you can buy more of these sure. Also im about to change the ammo so all of the above shit does not even matter

**Comment by nesisknesis** [comment]
wheres the changelog

File: code/modules/uplink/uplink_items.dm [hunk 8]
```diff
@@ -449,6 +449,17 @@ GLOBAL_LIST_INIT(uplink_items, subtypesof(/datum/uplink_item))
 	surplus = 0
 	include_modes = list(/datum/game_mode/nuclear)
 
+/datum/uplink_item/dangerous/minigun
+	name = "M-546 Osprey"
+	desc = "A fully-loaded minigun which packs a big punch. \
+			This deadly giant weapon has a massive 500-round magazine of devastating 5.46mm ammunition. Slaughter your enemies through \
```
Comment (nesisknesis) [inline code/modules/uplink/uplink_items.dm]:
```suggestion
			This deadly giant weapon has a massive 500-round magazine of devastating 5.46mm caseless ammunition. Slaughter your enemies through \
```

File: code/modules/uplink/uplink_items.dm [hunk 9]
```diff
@@ -449,6 +449,17 @@ GLOBAL_LIST_INIT(uplink_items, subtypesof(/datum/uplink_item))
...
+			sheer force and dont worry about bullet casings as the minigun shoots the whole bullet which is 65% more bullet per bullet. "
```
Comment (nesisknesis) [inline code/modules/uplink/uplink_items.dm]:
```suggestion
			sheer force."
```

**Review by nesisknesis — COMMENTED** [review]
you'll be crucified by gun experts for this

**Comment by Firewolf34** [comment]
I just wanted to say that the Portal reference in the description for this is hilarious, and very fitting given the circumstances.

**Comment by AsV9** [comment]
why does this have 31 commits

**Comment by Readystorm546** [comment]
> why does this have 31 commits

Balance

**Comment by Firewolf34** [comment]
> why does this have 31 commits

This is no problem - maintainer just has to `git merge --squash` when it's ready if it's really that much of a concern.

Better to have many commits than one really big one

Also @Readystorm546 : "We mad this gun soadvanced that it fires the whole bullet"

**Comment by Readystorm546** [comment]
> > why does this have 31 commits
> 
> This is no problem - maintainer just has to `git merge --squash` when it's ready if it's really that much of a concern.
> 
> Better to have many commits than one really big one
> 
> Also @Readystorm546 : "We mad this gun soadvanced that it fires the whole bullet"

Me and my spelling. I will fix it

**Comment by alexkar598** [comment]
This is going to need an actual description of what it does

File: yogstation/code/modules/projectiles/guns/ballistic/minigun.dm [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+	if(ammo_pack)
+		if(ammo_pack.overheat < ammo_pack.overheat_max)
+			ammo_pack.overheat += burst_size
+			..()
+		else
+			to_chat(user, "The gun's heat sensor locked the trigger to prevent heat damage.")
+
+/obj/item/gun/ballistic/minigunosprey/afterattack(atom/target, mob/living/user, flag, params)
+	if(!ammo_pack || ammo_pack.loc != user)
+		to_chat(user, "You need the more ammo to fire the gun!")
```
Comment (Bibby0110) [inline yogstation/code/modules/projectiles/guns/ballistic/minigun.dm:162]: You need the more ammo to fire the gun!

**Pull request closed without merging**